tools/state-viewer/src/contract_accounts.rs

```diff
@@ -0,0 +1,164 @@
+//! State viewer functions to list and filter accounts that have contracts
+//! deployed.
+
+use near_primitives::hash::CryptoHash;
+use near_primitives::trie_key::trie_key_parsers::parse_account_id_from_contract_code_key;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::AccountId;
+use near_store::{NibbleSlice, StorageError, Trie, TrieTraversalItem};
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+/// Output type for contract account queries with all relevant data around a
+/// single contract.
+pub(crate) struct ContractAccount {
+    pub(crate) account_id: AccountId,
+    pub(crate) source_wasm: Arc<[u8]>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ContractAccountError {
+    #[error("could not parse key {1:?}")]
+    InvalidKey(#[source] std::io::Error, Vec<u8>),
+    #[error("failed loading contract code for account {1}")]
+    NoCode(#[source] StorageError, AccountId),
+}
+
+impl std::fmt::Display for ContractAccount {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:<64} {:>9}", self.account_id, self.source_wasm.len())
+    }
+}
+
+impl ContractAccount {
+    /// Iterate over all contracts stored in the given trie, in lexicographic
+    /// order of the account IDs.
+    pub(crate) fn in_trie(trie: &Trie) -> anyhow::Result<ContractAccountIterator> {
+        ContractAccountIterator::new(trie)
+    }
+
+    fn from_contract_trie_node(
+        trie_key: &[u8],
+        value_hash: CryptoHash,
+        trie: &Trie,
+    ) -> Result<Self, ContractAccountError> {
+        let account_id = parse_account_id_from_contract_code_key(trie_key)
+            .map_err(|err| ContractAccountError::InvalidKey(err, trie_key.to_vec()))?;
+        let source_wasm = trie
+            .storage
+            .retrieve_raw_bytes(&value_hash)
+            .map_err(|err| ContractAccountError::NoCode(err, account_id.clone()))?;
+        Ok(Self { account_id, source_wasm })
+    }
+}
+
+pub(crate) struct ContractAccountIterator<'a> {
+    /// Trie nodes that point to the contracts.
+    contract_nodes: VecDeque<TrieTraversalItem>,
+    trie: &'a Trie,
+}
+
+impl<'a> ContractAccountIterator<'a> {
+    pub(crate) fn new(trie: &'a Trie) -> anyhow::Result<Self> {
+        let mut trie_iter = trie.iter()?;
+        // `TrieKey::ContractCode` requires a valid `AccountId`, we use "xx"
+        let key = TrieKey::ContractCode { account_id: "xx".parse()? }.to_vec();
+        let (prefix, suffix) = key.split_at(key.len() - 2);
+        assert_eq!(suffix, "xx".as_bytes());
+
+        // `visit_nodes_interval` wants nibbles stored in `Vec<u8>` as input
+        let nibbles_before: Vec<u8> = NibbleSlice::new(prefix).iter().collect();
+        let nibbles_after = {
+            let mut tmp = nibbles_before.clone();
+            *tmp.last_mut().unwrap() += 1;
+            tmp
+        };
```

Comment:
just fyi: I'm working on exposing API around trie key ranges as part of #8332, I will try to keep this use case in mind when designing it, so we can replace this with something nicer.

Reply:
good to know, thanks